registry/etcd/etcd.go

```diff
@@ -143,9 +154,26 @@ func (e *etcdRegistry) registerNode(s *registry.Service, node *registry.Node, op
 		return errors.New("Require at least one node")
 	}
 
-	// check existing lease cache
+	// parse the options, fallback to default domain
+	var options registry.RegisterOptions
+	for _, o := range opts {
+		o(&options)
+	}
+	if len(options.Domain) == 0 {
+		options.Domain = e.options.Domain
+	}
+
 	e.RLock()
-	leaseID, ok := e.leases[s.Name+node.Id]
+	// ensure the leases and registers are setup for this domain
+	if _, ok := e.leases[options.Domain]; !ok {
+		e.leases[options.Domain] = make(leases)
```

Comment:
You're going to need a write lock not a read lock

Reply:
Thanks. I'll amend.